### PR TITLE
Add new standalone file-browser in Lua

### PR DESCRIPTION
**Pull request by kilbith**
New standalone file-browser made in Lua, extensible to sort by file extension.

_Don't pay attention to the buttons colors, it's Kidscode defaults only_
The tree path is clickable by level using `hypertext`
Capture d’écran de 2020-10-31 19-25-05

Irrlicht file browser for comparison
Capture d’écran de 2020-10-31 19-22-26

### CODE REVIEW

**Comment by appgurueu** [comment]
Beautiful like all your interfaces. Love it.

**Comment by paramat** [comment]
Thanks for the feature, but please post new screenshots that do not contain and highlight 'fsck_paramat.png' files. I have removed your screenshots.

**Comment by MoNTE48** [comment]
Oh, how touchy we are. And also how we love to edit other people's messages. ^^^

File: builtin/mainmenu/dlg_file_browser.lua [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+-- This program is distributed in the hope that it will be useful,
+-- but WITHOUT ANY WARRANTY; without even the implied warranty of
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU Lesser General Public License for more details.
+--
+-- You should have received a copy of the GNU Lesser General Public License along
+-- with this program; if not, write to the Free Software Foundation, Inc.,
+-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+local ESC = core.formspec_escape
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_file_browser.lua:19]: Seeing as `ESC` is only used for textures, how about:
```lua
local function texture(filename) return core.formspec_escape(defaulttexturedir .. filename .. ".png") end
```

File: builtin/mainmenu/dlg_file_browser.lua [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	for dirname in string.gmatch(PATH, "([^" .. DIR_DELIM .. "]+)") do
+		value = value .. DIR_DELIM:gsub("\\", "\\\\") .. (dirname:gsub("_", "_u"):gsub(" ", "_s"))
+		_path = ("%s %s <action name=%s>%s</action>")
+			:format(_path, DIR_DELIM:gsub("\\", "\\\\\\\\"), value, dirname)
+	end
+
+	local fs = "size[10,7]" ..
+		"real_coordinates[true]" ..
+		"image_button[0.2,0.15;0.5,0.5;" ..
+			ESC(defaulttexturedir .. "arrow_up.png") .. ";updir;]" ..
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_file_browser.lua:69]: This for instance would accordingly be `texture("arrow_up")`. This entire formspec however should IMO be a string to be formatted using `string.format`.

File: builtin/mainmenu/dlg_file_browser.lua [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	end
+
+	for _, f in ipairs(dirs) do
+		local is_file = f:match("^.+(%..+)$")
+		_dirs = _dirs .. (is_file and "1," or "0,") .. f .. ","
+
+		if is_file then
+			local size = filesize(PATH .. DIR_DELIM .. f)
+			if size ~= nil then
+				size = size / 1000
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_file_browser.lua:98]: Filesize code can be shortened and beautified. Use something like:
```lua
local index = math.floor(math.log10(size) / 3)
local unit = ({"", "K", "M", "G", "T"})[index]
_dirs = _dirs .. string.format("%.1f", size / math.pow(10, index * 3)) .. " " .. unit .. "B,"
```

File: builtin/mainmenu/dlg_file_browser.lua [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		end
+	end
+
+	fs = fs .. "table[0.2,0.8;9.6,4.4;dirs;" .. _dirs:sub(1,-2) ..
+		";" .. (tabdata.selected or 1) .. "]"
+
+	return fs
+end
+
+local function fields_handler(this, fields)
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_file_browser.lua:121]: It's usually called `self`...

File: builtin/mainmenu/dlg_file_browser.lua [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		local event, idx = fields.dirs:sub(1,3), tonumber(fields.dirs:match("%d+"))
+		local filename = dirs[idx]
+		local is_file = (filename ~= nil) and filename:find("^.+(%..+)$")
+
+		if event == "CHG" then
+			tabdata.filename = is_file and filename or ""
+			tabdata.selected = idx
+
+			core.update_formspec(this:get_formspec())
+			return true
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_file_browser.lua:143]: All these `return ... elseif`s with the newlines... Couldn't you remove the elses?

**Review by appgurueu — CHANGES_REQUESTED** [review]

File: builtin/mainmenu/dlg_file_browser.lua [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	if PATH == "\\" then -- Windows root dir
+		dirs = minetest.get_dir_list(os.getenv("HOMEDRIVE"))
+	else
+		dirs = minetest.get_dir_list(PATH)
+	end
+
+	local new = {}
+	for _, f in ipairs(dirs) do
+		if f:sub(1,1) ~= "." then
+			local is_file = f:match("^.+(%..+)$")
```
[the remark this answers is not in the document]
Comment (kilbith) [reply builtin/mainmenu/dlg_file_browser.lua:44]: I'm against using the `lfs` lib just for that

[the remark this answers is not in the document]
Comment (sfan5) [reply builtin/mainmenu/dlg_file_browser.lua:44]: Yeah, if extra functionality is needed write a C function and expose it to Lua. We won't be adding an entire library.

File: builtin/mainmenu/dlg_file_browser.lua [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+--
+-- You should have received a copy of the GNU Lesser General Public License along
+-- with this program; if not, write to the Free Software Foundation, Inc.,
+-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+local ESC = core.formspec_escape
+local PATH = os.getenv("HOME") or os.getenv("HOMEPATH") or core.get_worldpath()
+local tabdata = {}
+
+local function filesize(name)
```
[the remark this answers is not in the document]
Comment (sfan5) [reply builtin/mainmenu/dlg_file_browser.lua:23]: There is not much wrong with this approach, though a `stat` call would be more efficient (and could also solve the folder or not issue below).

**Comment by appgurueu** [comment]
A secured `lfs` would definitely be a cleaner solution than reinventing the wheel.

**Comment by celeron55** [comment]
I'm closing this PR because of the aggressive language used by kilbith.

kilbith can re-create this PR without the aggressive language if he wishes this PR to exist.

If kilbith comes back with aggressive language, I will reinstantiate his permanent ban.

**Comment by lhofhansl** [comment]
It's a bit of a shame. I like this feature, at the same time we do need aggressive language, especially when it does not contribute to the discussion. For some it might be joke, and for some it does not feel this way. I like spirited discussion as long as it is constructive.

@kilbith are you willing to file a new PR with the same code? It's certainly the most effective way to influence the project in the direction you like. I'll happily review then.

**Comment by kilbith** [comment]
No I’m not willing to re-open it. I give once, not twice

**Comment by lhofhansl** [comment]
¯\_(ツ)_/¯

File: builtin/mainmenu/dlg_file_browser.lua (hunk 7 above)
[the remark this answers is not in the document]
Comment (Kimapr) [reply builtin/mainmenu/dlg_file_browser.lua:23]: > Reading a file just for getting it's size? Please involve `lfs` instead.

But it doesn't read the file, just opens it and seeks to the end.

**Pull request closed without merging**